gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemBase.java

```diff
@@ -819,6 +825,37 @@ public FSDataOutputStream append(Path hadoopPath, int bufferSize, Progressable p
     throw new IOException("The append operation is not supported.");
   }
 
+  /**
+   * Concat existing files together.
+   * @param trg the path to the target destination.
+   * @param psrcs the paths to the sources to use for the concatenation.
+   * @throws IOException IO failure
+   */
+  @Override
+  public void concat(final Path trg, final Path [] psrcs) throws IOException {
+    logger.atFine().log("GHFS.concat: %s, %d sources", trg, psrcs.length);
+
+    Preconditions.checkArgument(psrcs.length > 0, "must be at least one source");
+
+    URI trgPath = getGcsPath(trg);
+    List<URI> srcPaths = Arrays.stream(psrcs).map(this::getGcsPath).collect(Collectors.toList());
+
+    Preconditions.checkArgument(!srcPaths.contains(trgPath), "target must not be contained in sources");
+
+    List<List<URI>> partitions = Lists.partition(srcPaths, MAX_COMPOSE_OBJECTS - 1);
+    logger.atFine().log("GHFS.concat: %s, %d partitions", trg, partitions.size());
+    for (List<URI> partition : partitions) {
+      // We need to include the target in the list of sources to compose since
+      // the GCSFS compose operation will overwrite the target, whereas the Hadoop
+      // concat operation appends to the target.
+      List<URI> sources = Lists.newArrayList(trgPath);
```

Comment:
What will happen if target object does not exist but it's included in first `sources` partition? Does Hadoop concat semantic requires that target file should exist?

Reply:
The operation will fail in that case because target files doesn't exist. The target is required to exist - even if empty - for Hadoop FS concat to succeed.